### DISCUSSION TITLE
useMutation's reset() method runs on every other call from inside then() callback

### DISCUSSION THREAD

**Discussion by prajotKeywordio** [reduxjs/redux-toolkit#3326] [Q&A] [answered]
Hello everyone

I wanted to try and reset a mutation to its original state after success. I added the reset() function inside the then() callback of the mutation trigger's returned promise.

```
export const CreatePost = () => {
  const [createPost, { reset }] = useCreatePostMutation();

  const handleClick = () => {
    createPost().then(() => reset());
  };

  return (
    <>
      <button onClick={handleClick}>simulate post creation</button>
    </>
  );
};
````

However, the hook's state is not being reset after every click, instead it's being reset after every alternate click.
I checked the Redux DevTools inspector and it shows that the removeMutationResult action is not sent on the first click and sent twice on the second click once with the first request's id and once with the second request's id.

![Screenshot 2023-04-05 131412 (2)](https://user-images.githubusercontent.com/102961769/230016023-27879573-4a8f-4403-bf42-5f4687287810.png)

I am new to RTK-query and unable to figure this out. Can someone help me find out the reason for this behaviour?
Thanks!

[codesandbox](https://codesandbox.io/s/hungry-wave-scszy8)

**Comment by phryneas** [comment, marked as answer]
The problem is that you will only get the `reset` method for the currently running request *in the next render*.

`createPost().then(() => reset());`

`reset` here is a reference to the `reset` function that existed before `createPost` was started.

You could for example work around it like this:

```js
  const [createPost, { reset }] = useCreatePostMutation();

  const resetRef = useRef(reset)
  resetRef.current = reset

  const handleClick = () => {
    createPost().then(() => resetRef.current());
  };
```

**Reply by prajotKeywordio** [reply]
Thanks!! This finally helped me understand the behaviour.

**Reply by shreyashByteridge** [reply]
This helps. Thanks!!
